lib/sbi/sbi_sse.c

```diff
@@ -0,0 +1,963 @@
+{
+	if (e->state != SSE_STATE_RUNNING)
+		return SBI_EINVALID_STATE;
+
+	if (e->hartid != current_hartid())
+		return SBI_EDENIED;
+
+	if (flags & SBI_SSE_COMPLETE_FLAG_EVENT_DISABLE)
+		sse_event_disable(e);
+	else
+		sse_event_set_state(e, SSE_STATE_ENABLED);
+
+	if (e->cb_ops && e->cb_ops->complete_cb)
+		e->cb_ops->complete_cb(e->event_id);
+
+	if (sse_process_pending_events(e, regs) == SBI_EJUMP)
+		return SBI_EJUMP;
+
+	return sse_event_resume(e, regs);
+}
+
+int sbi_sse_complete(uint32_t event_id, uint32_t status, uint32_t flags,
+		     struct sbi_trap_regs *regs)
+{
+	int ret;
+	struct sbi_sse_event *e;
+
+	e = sse_event_get(event_id);
+	if (!e)
+		return SBI_EINVAL;
+
+	sse_event_lock(e);
+	ret = sse_event_complete(e, status, flags, regs);
+	sse_event_unlock(e);
+
+	return ret;
+}
+
+int sbi_sse_enable(uint32_t event_id)
+{
+	int ret;
+	struct sbi_sse_event *e;
+
+	e = sse_event_get(event_id);
+	if (!e)
+		return SBI_EINVAL;
+
+	sse_event_lock(e);
+	ret = sse_event_enable(e);
+	sse_event_unlock(e);
+
+	return ret;
+}
+
+int sbi_sse_disable(uint32_t event_id)
+{
+	int ret;
+	struct sbi_sse_event *e;
+
+	e = sse_event_get(event_id);
+	if (!e)
+		return SBI_EINVAL;
+
+	sse_event_lock(e);
+	ret = sse_event_disable(e);
+	sse_event_unlock(e);
+
+	return ret;
+}
+
+int sbi_sse_inject_from_ecall(uint32_t event_id, unsigned long hartid,
+			      struct sbi_trap_regs *regs)
+{
+	if (!sbi_domain_is_assigned_hart(sbi_domain_thishart_ptr(), hartid))
+		return SBI_EINVAL;
+
+	return sse_inject_event(event_id, hartid, regs, true);
+}
+
+int sbi_sse_inject_event(uint32_t event_id, struct sbi_trap_regs *regs)
+{
+	return sse_inject_event(event_id, current_hartid(), regs, false);
+}
+
+int sbi_sse_set_cb_ops(uint32_t event_id, const struct sbi_sse_cb_ops *cb_ops)
+{
+	struct sbi_sse_event *e;
+
+	e = sse_event_get(event_id);
+	if (!e)
+		return SBI_EINVAL;
+
+	if (cb_ops->set_hartid_cb && !sse_event_is_global(e))
+		return SBI_EINVAL;
+
+	sse_event_lock(e);
+	e->cb_ops = cb_ops;
+	sse_event_unlock(e);
+
+	return SBI_OK;
+}
+
+
+int sbi_sse_get_attr(uint32_t event_id, uint32_t attr_id, unsigned long *out_val)
+{
+	struct sbi_sse_event *e;
+
+	e = sse_event_get(event_id);
+	if (!e)
+		return SBI_EINVAL;
+
+	return sse_event_get_attr(e, attr_id, out_val);
+}
+
+int sbi_sse_set_attr(uint32_t event_id, uint32_t attr_id, unsigned long val)
+{
+	int ret;
+	struct sbi_sse_event *e;
+
+	e = sse_event_get(event_id);
+	if (!e)
+		return SBI_EINVAL;
+
+	sse_event_lock(e);
+	ret = sse_event_set_attr(e, event_id, attr_id, val);
+	sse_event_unlock(e);
+
+	return ret;
+}
+
+int sbi_sse_register(uint32_t event_id,
+			unsigned long phys_lo,
+			unsigned long phys_hi)
+{
+	int ret;
+	struct sbi_sse_event *e;
+	const unsigned align = __riscv_xlen >> 3;
+	ulong smode = (csr_read(CSR_MSTATUS) & MSTATUS_MPP) >>
+			MSTATUS_MPP_SHIFT;
+
+	if (phys_lo & (align - 1))
+		return SBI_EINVALID_ADDR;
+
+	/*
+	 * On RV32, the M-mode can only access the first 4GB of
```

Comment:
Shouldn't be an issue for RV64.
This gets compiled for RV64 too, right?

So `phys_hi` check should be under some `ifdef`, right?
Am I missing something here?